File: src/components/blog/BlogAiTools.js
import React, { useState, useCallback} from 'react';
import {
    Typography,
    CircularProgress,
    AccordionSummary,
    AccordionDetails,
    MenuItem,
    Select,
    FormControl,
    InputLabel,
    Box,
    FormControlLabel,
    Switch,
} from '@mui/material';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import blogAiService from '../../utils/blogAiService';
import {
    AiToolsContainer,
    AiToolsCard,
    AiToolsAccordion,
    AiToolsButton,
    AiToolsForm,
    AiToolsPreview,
    AiToolsLoadingContainer,
    AiToolsResultContainer,
} from '../styles/BlogAiTools.styles';

// Custom hook for managing AI tool operations
const useAiTool = (initialContent) => {
    const [loading, setLoading] = useState(false);
    const [selectedText, setSelectedText] = useState('');
    const [useSelectedTextOnly, setUseSelectedTextOnly] = useState(true);
    const [targetAudience, setTargetAudience] = useState('general');
    const [tone, setTone] = useState('professional');
    const [results, setResults] = useState({});

    // Get the text to process (either selected text or full content)
    const textToProcess = useCallback(() => {
        if (useSelectedTextOnly && selectedText) {
            return selectedText;
        }
        return initialContent;
    }, [initialContent, selectedText, useSelectedTextOnly]);

    // Handle text selection
    const handleTextSelection = useCallback(() => {
        const selection = window.getSelection();
        if (selection.toString()) {
            setSelectedText(selection.toString());
        }
    }, []);

    // Generic function to call any AI service
    const callAiService = useCallback(
        async (serviceFunction, params, resultKey) => {
            setLoading(true);
            try {
                const result = await serviceFunction(...params);
                setResults((prev) => ({ ...prev, [resultKey]: result }));
                return result;
            } catch (error) {
                console.error(`Error calling ${resultKey}:`, error);
            } finally {
                setLoading(false);
            }
        },
        []
    );

    // Clear a specific result
    const clearResult = useCallback((key) => {
        setResults((prev) => {
            const newResults = { ...prev };
            delete newResults[key];
            return newResults;
        });
    }, []);

    // Clear all results
    const clearAllResults = useCallback(() => {
        setResults({});
    }, []);

    return {
        loading,
        selectedText,
        setSelectedText,
        useSelectedTextOnly,
        setUseSelectedTextOnly,
        targetAudience,
        setTargetAudience,
        tone,
        setTone,
        results,
        textToProcess,
        handleTextSelection,
        callAiService,
        clearResult,
        clearAllResults,
    };
};

// Tool option component
const ToolOption = ({ title, onExecute, resultKey, result, onApply }) => {
    return (
        <AiToolsAccordion>
            <AccordionSummary expandIcon={<ExpandMoreIcon />}>
                <Typography>{title}</Typography>
            </AccordionSummary>
            <AccordionDetails>
                <AiToolsButton
                    variant="contained"
                    onClick={onExecute}
                    fullWidth
                >
                    Execute
                </AiToolsButton>
                {result && (
                    <AiToolsResultContainer>
                        <Typography variant="h6" gutterBottom>
                            Results
                        </Typography>
                        <AiToolsPreview>
                            <pre>{JSON.stringify(result, null, 2)}</pre>
                        </AiToolsPreview>
                        <AiToolsButton
                            variant="contained"
                            color="primary"
                            onClick={() => onApply(result)}
                            sx={{ mt: 2 }}
                        >
                            Apply Changes
                        </AiToolsButton>
                    </AiToolsResultContainer>
                )}
            </AccordionDetails>
        </AiToolsAccordion>
    );
};

// Selection preview component
const SelectionPreview = ({
    selectedText,
    useSelectedTextOnly,
    onToggleUseSelected,
}) => {
    if (!selectedText) return null;

    return (
        <Box sx={{ mt: 2, mb: 2 }}>
            <Typography variant="subtitle1" gutterBottom>
                Selected Text:
            </Typography>
            <AiToolsPreview>
                <Typography variant="body2">{selectedText}</Typography>
            </AiToolsPreview>
            <FormControlLabel
                control={
                    <Switch
                        checked={useSelectedTextOnly}
                        onChange={(e) => onToggleUseSelected(e.target.checked)}
                    />
                }
                label="Use selected text only"
                sx={{ mt: 1 }}
            />
        </Box>
    );
};

// Main component
const BlogAiTools = ({ content, onApplyChanges }) => {
    const {
        loading,
        selectedText,
        useSelectedTextOnly,
        setUseSelectedTextOnly,
        targetAudience,
        setTargetAudience,
        tone,
        setTone,
        results,
        textToProcess,
        handleTextSelection,
        callAiService,
    } = useAiTool(content);

    // Handler functions for each AI tool
    const handleOrganizeThoughts = useCallback(() => {
        callAiService(
            blogAiService.organizeThoughts,
            [textToProcess()],
            'organizedThoughts'
        );
    }, [callAiService, textToProcess]);

    const handleEditContent = useCallback(() => {
        callAiService(
            blogAiService.editContent,
            [textToProcess(), targetAudience, tone],
            'editedContent'
        );
    }, [callAiService, textToProcess, targetAudience, tone]);

    const handleGenerateTitles = useCallback(() => {
        callAiService(
            blogAiService.generateTitles,
            [textToProcess(), textToProcess().split('\n')[0]],
            'titles'
        );
    }, [callAiService, textToProcess]);

    const handleExpandPoints = useCallback(() => {
        callAiService(
            blogAiService.expandBriefPoints,
            [textToProcess(), tone],
            'expandedPoints'
        );
    }, [callAiService, textToProcess, tone]);

    const handleResearchDirections = useCallback(() => {
        callAiService(
            blogAiService.generateResearchDirections,
            [textToProcess().split('\n')[0]],
            'researchDirections'
        );
    }, [callAiService, textToProcess]);

    const handleAdjustTone = useCallback(() => {
        callAiService(
            blogAiService.adjustTone,
            [textToProcess(), tone],
            'adjustedTone'
        );
    }, [callAiService, textToProcess, tone]);

    const handleGenerateConclusion = useCallback(() => {
        callAiService(
            blogAiService.generateConclusion,
            [textToProcess()],
            'conclusion'
        );
    }, [callAiService, textToProcess]);

    const handleGenerateIntroduction = useCallback(() => {
        callAiService(
            blogAiService.generateIntroduction,
            [textToProcess().split('\n')[0], targetAudience],
            'introduction'
        );
    }, [callAiService, textToProcess, targetAudience]);

    return (
        <AiToolsContainer onMouseUp={handleTextSelection}>
            <AiToolsCard>
                <Typography variant="h5" gutterBottom>
                    AI Writing Assistant
                </Typography>

                <SelectionPreview
                    selectedText={selectedText}
                    useSelectedTextOnly={useSelectedTextOnly}
                    onToggleUseSelected={setUseSelectedTextOnly}
                />

                <AiToolsForm>
                    <FormControl fullWidth>
                        <InputLabel>Target Audience</InputLabel>
                        <Select
                            value={targetAudience}
                            onChange={(e) => setTargetAudience(e.target.value)}
                            label="Target Audience"
                        >
                            <MenuItem value="general">General</MenuItem>
                            <MenuItem value="technical">Technical</MenuItem>
                            <MenuItem value="business">Business</MenuItem>
                            <MenuItem value="academic">Academic</MenuItem>
                        </Select>
                    </FormControl>

                    <FormControl fullWidth>
                        <InputLabel>Tone</InputLabel>
                        <Select
                            value={tone}
                            onChange={(e) => setTone(e.target.value)}
                            label="Tone"
                        >
                            <MenuItem value="professional">
                                Professional
                            </MenuItem>
                            <MenuItem value="casual">Casual</MenuItem>
                            <MenuItem value="formal">Formal</MenuItem>
                            <MenuItem value="friendly">Friendly</MenuItem>
                            <MenuItem value="authoritative">
                                Authoritative
                            </MenuItem>
                        </Select>
                    </FormControl>

                    {loading ? (
                        <AiToolsLoadingContainer>
                            <CircularProgress />
                        </AiToolsLoadingContainer>
                    ) : (
                        <>
                            <ToolOption
                                title="Organize Thoughts"
                                onExecute={handleOrganizeThoughts}
                                result={results.organizedThoughts}
                                onApply={onApplyChanges}
                            />

                            <ToolOption
                                title="Edit Content"
                                onExecute={handleEditContent}
                                result={results.editedContent}
                                onApply={onApplyChanges}
                            />

                            <ToolOption
                                title="Generate Titles"
                                onExecute={handleGenerateTitles}
                                result={results.titles}
                                onApply={onApplyChanges}
                            />

                            <ToolOption
                                title="Expand Points"
                                onExecute={handleExpandPoints}
                                result={results.expandedPoints}
                                onApply={onApplyChanges}
                            />

                            <ToolOption
                                title="Research Directions"
                                onExecute={handleResearchDirections}
                                result={results.researchDirections}
                                onApply={onApplyChanges}
                            />

                            <ToolOption
                                title="Adjust Tone"
                                onExecute={handleAdjustTone}
                                result={results.adjustedTone}
                                onApply={onApplyChanges}
                            />

                            <ToolOption
                                title="Generate Conclusion"
                                onExecute={handleGenerateConclusion}
                                result={results.conclusion}
                                onApply={onApplyChanges}
                            />

                            <ToolOption
                                title="Generate Introduction"
                                onExecute={handleGenerateIntroduction}
                                result={results.introduction}
                                onApply={onApplyChanges}
                            />
                        </>
                    )}
                </AiToolsForm>
            </AiToolsCard>
        </AiToolsContainer>
    );
};

export default BlogAiTools;

File: server/routes/blog_ai_tools.py
from fastapi import APIRouter, Depends, Request
from server.utils.ai_helpers import try_process_ai_request
from server.models.blog import BlogContentOnly, BlogContentWithTone
from server.agents.blog.agent_blog import (
    organize_thoughts,
    edit_blog_content,
    generate_titles,
    expand_blog_content,
    generate_research_directions,
    generate_conclusion,
    generate_introduction,
    adjust_tone,
    prepare_publishing_package
)

router = APIRouter(prefix="/api/blog/ai", tags=["blog-ai"])

async def get_blog_collection(request: Request):
    return request.app.state.mongo_client.get_blog_posts_collection()

@router.post("/organize-thoughts")
async def organize(request: BlogContentOnly, collection=Depends(get_blog_collection)):
    return await try_process_ai_request(
        organize_thoughts,
        result_key="organizedThoughts",
        collection=collection,
        post_id=request.post_id,
        raw_thoughts=request.blog_content  # <- Function-specific kwarg
    )

@router.post("/edit-content")
async def edit(request: BlogContentWithTone, collection=Depends(get_blog_collection)):
    return await try_process_ai_request(
        edit_blog_content,
        result_key="editedContent",
        collection=collection,
        post_id=request.post_id,
        blog_content=request.blog_content,
        tone=request.tone
    )

@router.post("/generate-titles")
async def titles(request: BlogContentWithTone, collection=Depends(get_blog_collection)):
    return await try_process_ai_request(
        generate_titles,
        result_key="titles",
        collection=collection,
        post_id=request.post_id,
        blog_content=request.blog_content,
        tone=request.tone
    )


@router.post("/expand-blog-content")
async def expand(request: BlogContentWithTone, collection=Depends(get_blog_collection)):
    return await try_process_ai_request(
        expand_blog_content,
        result_key="expandedPoints",
        collection=collection,
        post_id=request.post_id,
        blog_content=request.blog_content,
        tone=request.tone
    )

@router.post("/generate-research-directions")
async def research(request: BlogContentOnly, collection=Depends(get_blog_collection)):
    return await try_process_ai_request(
        generate_research_directions,
        result_key="researchDirections",
        collection=collection,
        post_id=request.post_id,
        blog_content=request.blog_content
    )


@router.post("/adjust-tone")
async def tone(request: BlogContentWithTone, collection=Depends(get_blog_collection)):
    return await try_process_ai_request(
        adjust_tone,
        result_key="adjustedTone",
        collection=collection,
        post_id=request.post_id,
        blog_content=request.blog_content,
        tone=request.tone
    )


@router.post("/generate-conclusion")
async def conclusion(request: BlogContentWithTone, collection=Depends(get_blog_collection)):
    return await try_process_ai_request(
        generate_conclusion,
        result_key="conclusion",
        collection=collection,
        post_id=request.post_id,
        blog_content=request.blog_content,
        tone=request.tone
    )

@router.post("/generate-introduction")
async def intro(request: BlogContentWithTone, collection=Depends(get_blog_collection)):
    return await try_process_ai_request(
        generate_introduction,
        result_key="introduction",
        collection=collection,
        post_id=request.post_id,
        blog_content=request.blog_content,
        tone=request.tone
    )

@router.post("/prepare-publishing-package")
async def publishing(request: BlogContentWithTone, collection=Depends(get_blog_collection)):
    return await try_process_ai_request(
        prepare_publishing_package,
        result_key="publishingPackage",
        collection=collection,
        post_id=request.post_id,
        blog_content=request.blog_content,
        tone=request.tone
    )